packages/x-data-grid-pro/src/components/GridServerSideTreeDataGroupingCell.tsx

```diff
@@ -0,0 +1,220 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import { unstable_composeClasses as composeClasses } from '@mui/utils';
+import { styled } from '@mui/system';
+import Box from '@mui/material/Box';
+import Badge from '@mui/material/Badge';
+import {
+  getDataGridUtilityClass,
+  GridRenderCellParams,
+  GridServerSideGroupNode,
+  useGridSelector,
+} from '@mui/x-data-grid';
+import CircularProgress from '@mui/material/CircularProgress';
+import { useGridRootProps } from '../hooks/utils/useGridRootProps';
+import { useGridPrivateApiContext } from '../hooks/utils/useGridPrivateApiContext';
+import { DataGridProProcessedProps } from '../models/dataGridProProps';
+import { GridPrivateApiPro } from '../models/gridApiPro';
+import { GridStatePro } from '../models/gridStatePro';
+
+type OwnerState = { classes: DataGridProProcessedProps['classes'] };
+
+const useUtilityClasses = (ownerState: OwnerState) => {
+  const { classes } = ownerState;
+
+  const slots = {
+    root: ['treeDataGroupingCell'],
+    toggle: ['treeDataGroupingCellToggle'],
+  };
+
+  return composeClasses(slots, getDataGridUtilityClass, classes);
+};
+
+interface GridTreeDataGroupingCellProps
+  extends GridRenderCellParams<any, any, any, GridServerSideGroupNode> {
+  hideDescendantCount?: boolean;
+  /**
+   * The cell offset multiplier used for calculating cell offset (`rowNode.depth * offsetMultiplier` px).
+   * @default 2
+   */
+  offsetMultiplier?: number;
+}
+
+interface GridTreeDataGroupingCellIconProps
+  extends Pick<GridTreeDataGroupingCellProps, 'id' | 'field' | 'rowNode' | 'row'> {
+  descendantCount: number;
+}
+
+const LoadingContainer = styled('div')({
+  display: 'flex',
+  alignItems: 'center',
+  justifyContent: 'center',
+  height: '100%',
+});
+
+function GridTreeDataGroupingCellIcon(props: GridTreeDataGroupingCellIconProps) {
+  const apiRef = useGridPrivateApiContext() as React.MutableRefObject<GridPrivateApiPro>;
+  const rootProps = useGridRootProps();
+  const { rowNode, id, field, descendantCount } = props;
+
+  const loadingSelector = React.useCallback(
```

Comment:
Right, so the difference here would be when the `loading` state gets updated:
- `gridServerSideDataLoadingSelector`: every `GridTreeDataGroupingCellIcon` instance will rerender
- current approach: only impacted instances will rerender.

Correct?
It might be an overoptimization, but I see the value now.
Feel free to resolve this discussion and keep it as is.

Reply:
Just to note, there isn't any advantage to using `useCallback` here, it's actually worst to use it. `useGridSelector` doesn't reactively update when the selector changes (there wasn't a case where that made sense).

The pattern `useGridSelector(state => state.x.y)` might cause problems with the fine-grained reactivity refactor I proposed a few weeks ago, but we already have other cases like that so you can use it without problem right now.